service/openapi/src/main/java/org/dotwebstack/framework/service/openapi/response/ResponseContextHelper.java

```diff
@@ -44,44 +45,63 @@ private ResponseContextHelper() {}
-          .map(path -> {
-            if (path.startsWith(responseObject.getIdentifier() + ".")) {
-              return path.substring(path.indexOf('.') + 1);
-            }
-            return path;
-          })
-          .filter(path -> !path.isEmpty())
-          .collect(Collectors.toSet());
     }
+
+    return new HashSet<>(getRequiredResponseObject("", responseObject, graphQlField, inputParams).keySet());
   }
 
   static Map<String, SchemaSummary> getRequiredResponseObject(String prefix, ResponseObject responseObject,
-      GraphQlField graphQlField, Map<String, Object> inputParams, boolean skipPath) {
+      GraphQlField graphQlField, Map<String, Object> inputParams) {
     Map<String, SchemaSummary> responseObjects = new HashMap<>();
     StringJoiner joiner = getStringJoiner(prefix);
-
     SchemaSummary summary = responseObject.getSummary();
-    boolean skip = skipPath;
-    if (!summary.isEnvelope() && !Objects.equals(summary.getType(), OasConstants.ARRAY_TYPE) && summary.getComposedOf()
-        .isEmpty()) {
-      if (!skipPath || !Objects.equals(summary.getType(), OasConstants.OBJECT_TYPE)) {
-        joiner.add(responseObject.getIdentifier());
-        if (summary.isRequired()) {
-          responseObjects.put(joiner.toString(), summary);
-        }
-      }
-      skip = false;
-    }
+    addPrefixToPath(summary, responseObject, joiner, responseObjects);
     if (summary.isRequired() || summary.isEnvelope()
         || isExpanded(inputParams, getPathString(prefix, responseObject))) {
-      handleSubSchemas(graphQlField, inputParams, responseObjects, joiner, responseObject, skip);
+      handleSubSchemas(graphQlField, inputParams, responseObjects, joiner, responseObject);
     }
     return responseObjects;
   }
 
+  private static void addPrefixToPath(SchemaSummary summary, ResponseObject responseObject, StringJoiner joiner,
+      Map<String, SchemaSummary> responseObjects) {
+
+    if (summary.isEnvelope() || !summary.getComposedOf()
+        .isEmpty() || responseObject.getParent() == null || Objects.equals(OasConstants.ARRAY_TYPE, summary.getType())
+        || (Objects.equals(OasConstants.OBJECT_TYPE, summary.getType()) && hasDirectArrayParent(responseObject)
+            && hasAtMostOneArrayAncestor(responseObject))) {
+      return;
+    }
+
+    joiner.add(responseObject.getIdentifier());
+    if (summary.isRequired()) {
+      responseObjects.put(joiner.toString(), summary);
+    }
+  }
+
+  private static boolean hasDirectArrayParent(ResponseObject responseObject) {
+    return responseObject.getParent() != null && Objects.equals(ARRAY_TYPE, responseObject.getParent()
+        .getSummary()
+        .getType()) && responseObject.getIdentifier()
+            .equals(responseObject.getParent()
+                .getIdentifier());
+  }
+
+  private static boolean hasAtMostOneArrayAncestor(ResponseObject responseObject) {
+    ResponseObject parent = responseObject.getParent();
+    int listAncestors = 0;
```

Comment:
Ik zou niet ancestors en parents door elkaar gebruiken
```suggestion
    int listParents = 0;
```